llvm/lib/Transforms/IPO/FunctionAttrs.cpp

```diff
@@ -1760,6 +1761,70 @@ static void addNoReturnAttrs(const SCCNodeSet &SCCNodes,
   }
 }
 
+static bool
+allBBPathsGoThroughCold(BasicBlock *BB,
+                        SmallDenseMap<BasicBlock *, bool, 16> &Visited) {
+  // If BB contains a cold callsite this path through the CG is cold.
+  if (any_of(*BB, [](Instruction &I) {
+        if (auto *CB = dyn_cast<CallBase>(&I))
+          return CB->hasFnAttr(Attribute::Cold);
+        return false;
```

Comment:
I'd add a note that this assumes throwing an diverging code paths are cold, which is why this does not check for guaranteed-to-transfer.